### ISSUE TITLE
Fail to run shell demo

### ISSUE THREAD

**Issue by OscarShiang** [DavidSpickett/ARMMultiTasking#4]
**Describe the bug**
The shell demo fails to open stdin to read the input from terminal.

**Configuration**
 - cmake version (`cmake --version`)
```
cmake version 3.16.3
```
 - toolchain/compiler version (`gcc --version`, for example)
```
aarch64-none-elf-gcc (GNU Toolchain for the A-profile Architecture 9.2-2019.12 (arm-9.10)) 9.2.1 20191025
arm-none-eabi-gcc (GNU Arm Embedded Toolchain 9-2020-q2-update) 9.3.1 20200408 (release)
```
 - QEMU version (`qemu-system-arm --version`)
```
QEMU emulator version 5.1.0
QEMU emulator version 4.2.1
```

**To Reproduce (complete those that apply):**
- cmake configuration
```
cmake . -DBUILD_PLATFORM=arm
cmake . -DBUILD_PLATFORM=aarch64
cmake . -DBUILD_PLATFORM=thumb
```
- make command
```
make
```
- run command
```
make run_shell
```

**Expected behavior**

The shell demo should work properly. 

**Affected platforms/settings**
List any other settings/platforms/software versions you have tried and whether that had any effect.
e.g. "This issue presents at -O2 but not -O1."

This problem occurs all platflaform listed below
- `arm` 
- `aarch64`
- `thumb`

The error output is like this:

```shell
$ make run_shell
[ 57%] Built target shell
[ 71%] Built target ps
[ 85%] Built target echo
[100%] Built target ls
---------------------
----- AMT Shell -----
---------------------
$ /home/oscar/ARMMultiTasking/demos/shell/shell.c:236 (command_loop) Expected: (got != 0) && "stdin closed!"
make[3]: *** [CMakeFiles/run_shell.dir/build.make:57: run_shell] Error 1
make[2]: *** [CMakeFiles/Makefile2:894: CMakeFiles/run_shell.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:901: CMakeFiles/run_shell.dir/rule] Error 2
make: *** [Makefile:430: run_shell] Error 2
``` 

The problem may occurs in both version mentioned above, but in QEMU `2.11.2`, the shell demo works properly.
And I try to checkout to some previous commit, but it ends up to have some problems when I try to enter something

```shell
$ make run_shell
[ 57%] Built target shell
[ 71%] Built target ps
[ 85%] Built target echo
[100%] Built target ls
---------------------
----- AMT Shell -----
---------------------
$ d?�Um?�Uf?�UUnhandled escape sequence!
make[3]: *** [CMakeFiles/run_shell.dir/build.make:57: run_shell] Error 1
make[2]: *** [CMakeFiles/Makefile2:894: CMakeFiles/run_shell.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:901: CMakeFiles/run_shell.dir/rule] Error 2
make: *** [Makefile:430: run_shell] Error 2
``` 

It seems that QEMU had disabled some settings which is used to be turned on in `2.*` versions but I fail to find the proper setting of it.

Is there any clue to find or fix it?

**Comment by DavidSpickett** [comment]
Hi Oscar, yes this is the result of me trying to nail down how read() works over semihosting.

What I saw on QEMU 2.1 was if you read() N bytes then get N+1, that means nothing was read. So if you didn't type anything you'd just get 16 over and over.
I *assumed* that therefore 0 would mean the file was closed but I never managed to hit that assert.
This may have been an issue with QEMU or some undefined corner case in the spec. (sounds like an off by one to me)

Anyway, https://developer.arm.com/documentation/100863/0300/Semihosting-operations/SYS-READ--0x06-?lang=en is what the spec says.

I'd suggest removing the assert and changing this if:
https://github.com/DavidSpickett/ARMMultiTasking/blob/7ddee92be0ecf5fad5bac3e88ab354a9eb871562/demos/shell/shell.c#L238-L241

To be:
```
// Some comment about which qemu does what...
if (!got || got == INPUT_BUFFER_SIZE) {
      continue;
}
```

stdin being closed is unlikely so losing the assert isn't a big issue. If that works for you send a pull request, Azure will check qemu 2.1 if you don't already have it around.

Side note: There is a read from console semihosting call but I'm not sure of its behavior and I *think* read() should be enough here.
(https://developer.arm.com/documentation/100863/0300/Semihosting-operations/SYS-READC--0x07-?lang=en)
